[AST] Use cached deserialized decl in `getOpaqueResultTypeDecl`

hamishknight opened the pull request:
Mangling and looking up the opaque result type decl for serialized decls is a fairly expensive operation. Instead, fall through to the request which will have a cached value set by deserialization. This shaves ~30ms off the cached completion for:

```swift
import SwiftUI

struct V: View {
  var body: some View {
    Table(#^CC^#
  }
}
```

hamishknight commented:
@swift-ci please test

hamishknight commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -42,7 +42,9 @@ OpaqueTypeDecl *
                                   ValueDecl *originatingDecl) const {
   auto *repr = originatingDecl->getOpaqueResultTypeRepr();
-  assert(repr && "Declaration does not have an opaque result type");
+  if (!repr)
```
Currently deserialization only calls cacheOutput when an opaque decl is present, we could change that though

hamishknight replied:
Updated to cache the negative case too

slavapestov replied:
Can you convert OpaqueResultTypeRequest to use split caching so that we’re not populating a DenseMap with null values?

hamishknight commented:
@swift-ci please test

hamishknight committed:
[AST] Use cached deserialized decl in `getOpaqueResultTypeDecl`

Mangling and looking up the opaque result type decl
for serialized decls is a fairly expensive
operation. Instead, fallthrough to the request
which will have a cached value set by deserialization.
This shaves ~30ms off the cached completion for:

```swift
import SwiftUI

struct V: View {
  var body: some View {
    Table(#^CC^#
  }
}
```

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

slavapestov reviewed (approved):
Thanks!
